core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTypesOfCancers.java

```diff
@@ -32,74 +32,107 @@
-            //if not clobber, then existing cancer types should be skipped:
-            if (!clobber && DaoTypeOfCancer.getTypeOfCancerById(typeOfCancerId.toLowerCase()) != null ) {
-            	ProgressMonitor.logWarning("Cancer type with id '" + typeOfCancerId + "' already exists. Skipping.");
+        while (scanner.hasNextLine()) {
+            String nextLine = scanner.nextLine();
+            String[] fields = nextLine.split("\t", -1);
+            throwExceptionIfColumnCountIsWrong(file, nextLine, fields, EXPECTED_DATAFILE_COLUMN_COUNT);
+            TypeOfCancer typeOfCancer = new TypeOfCancer();
+            String typeOfCancerId = fields[0].trim();
+            typeOfCancer.setTypeOfCancerId(typeOfCancerId.toLowerCase());
+            typeOfCancer.setName(fields[1].trim());
+            typeOfCancer.setDedicatedColor(fields[2].trim());
+            typeOfCancer.setShortName(typeOfCancerId);
+            typeOfCancer.setParentTypeOfCancerId(fields[3].trim().toLowerCase());
+            typeOfCancerList.add(typeOfCancer);
+        }
+        return typeOfCancerList;
+    }
+
+    private static void throwExceptionIfColumnCountIsWrong(File file, String nextLine, String[] fields, int expectedDatafileColumnCount) throws IOException {
+        if (fields.length != expectedDatafileColumnCount) {
+            String msg = String.format("Cancer type file '%s' contains a line which is not %d-column tab-delimited." +
+                    " Expected fields can be seen here: https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#cancer-type" +
+                    " The invalid line has %d columns:\n%s", file.getPath(), expectedDatafileColumnCount, fields.length, nextLine);
+            for (int i = 0; i < fields.length ; i++) {
+                msg = msg + String.format("  field#%d(len%d):'%s'\n", i, fields[i].length(), fields[i]);
             }
-            else {
-	            aTypeOfCancer.setTypeOfCancerId(typeOfCancerId.toLowerCase());
-	            aTypeOfCancer.setName(tokens[1].trim());
-	            aTypeOfCancer.setDedicatedColor(tokens[2].trim());
-	            aTypeOfCancer.setShortName(typeOfCancerId);
-	            aTypeOfCancer.setParentTypeOfCancerId(tokens[3].trim().toLowerCase());
-	            DaoTypeOfCancer.addTypeOfCancer(aTypeOfCancer);
-	            numNewCancerTypes++;
+        throw new IOException(msg);
+        }
+    }
+
+    private static void writeRecordsToDatabase(List<TypeOfCancer> typeOfCancerList, boolean clobber) throws IOException, DaoException {
+        int numNewCancerTypes = 0;
+        for (TypeOfCancer typeOfCancer : typeOfCancerList) {
+            if (!clobber && typeOfCancerExistsInDatabase(typeOfCancer)) {
+                ProgressMonitor.logWarning(String.format("Cancer type with id '%s' already exists. Skipping.", typeOfCancer.getTypeOfCancerId()));
+            } else {
+                DaoTypeOfCancer.addTypeOfCancer(typeOfCancer);
```

Comment:
will an error be thrown if clobber = true and the type of cancer already exists?

Does an update occur to the existing entry?

Reply:
note: if clobber == true then the db table should already be wiped so this scenario shouldnt be possible?